GVFS/GVFS.Common/Git/GitSsl.cs

```diff
@@ -0,0 +1,171 @@
+            }
+
+            return result;
+        }
+
+        private static void LogCertificateCounts(ITracer tracer, EventMetadata metadata, IEnumerable<X509Certificate2> certificates, string messageTemplate)
+        {
+            Action<EventMetadata, string> loggingFunction;
+            int numberOfCertificates = certificates.Count();
+
+            switch (numberOfCertificates)
+            {
+                case 0:
+                    loggingFunction = tracer.RelatedError;
+                    break;
+                case 1:
+                    loggingFunction = tracer.RelatedInfo;
+                    break;
+                default:
+                    loggingFunction = tracer.RelatedWarning;
+                    break;
+            }
+
+            loggingFunction(
+                metadata,
+                string.Format(
+                    messageTemplate,
+                    numberOfCertificates,
+                    string.Join(
+                        Environment.NewLine,
+                        certificates.Select(x => x.Subject))));
+        }
+
+        private X509Certificate2 GetCertificateFromFile(ITracer tracer, EventMetadata metadata, string certificatePassword, bool onlyLoadValidCertificateFromStore)
+        {
+            if (File.Exists(this.CertificatePathOrSubjectCommonName))
+            {
+                try
+                {
+                    X509Certificate2 cert = new X509Certificate2(this.CertificatePathOrSubjectCommonName, certificatePassword);
+                    if (onlyLoadValidCertificateFromStore && cert != null && !cert.Verify())
+                    {
+                        tracer.RelatedWarning(metadata, "Certficate was found, but is invalid.");
+                        return null;
+                    }
+
+                    return cert;
+                }
+                catch (CryptographicException cryptEx)
+                {
+                    metadata.Add("Exception", cryptEx);
+                    tracer.RelatedError(metadata, "Error, while loading certificate from disk");
+                    return null;
+                }
+            }
+
+            return null;
+        }
+
+        private X509Certificate2 GetCertificateFromStore(ITracer tracer, EventMetadata metadata, bool onlyLoadValidCertificateFromStore)
+        {
+            X509Store store = null;
+            try
+            {
+                store = new X509Store();
+                store.Open(OpenFlags.OpenExistingOnly | OpenFlags.ReadOnly);
+
+                X509Certificate2Collection findResults = store.Certificates.Find(X509FindType.FindBySubjectName, this.CertificatePathOrSubjectCommonName, onlyLoadValidCertificateFromStore);
+                if (findResults?.Count > 0)
+                {
+                    LogCertificateCounts(tracer, metadata, findResults.OfType<X509Certificate2>(), "Found {0} certificates by provided name. Matching DNs: {1}");

```

Comment:
I don't think the string with format should be so far away from where it is used in case there is something that needs to change and one needs to pass more that 2 items for formatting.  Can we create the string here instead?
```
IEnumerable<X509Certificate2> certificates = findResults.OfType<X509Certificate2>();
string matchingNames = string.Join(Environment.NewLine, certificates.Select(x => x.Subject));
string logMessage = $"Found {certificates.Count()} certificates by provided name. Matching DNs: {matchingNames}";
```